README.md
# Amzscout-scrape

[![Version status](https://img.shields.io/pypi/status/amzscout-scrape)](https://pypi.org/project/amzscout-scrape)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![Python version compatibility](https://img.shields.io/pypi/pyversions/amzscout-scrape)](https://pypi.org/project/amzscout-scrape)
[![Version on Docker Hub](https://img.shields.io/docker/v/regulad/amzscout-scrape?color=green&label=Docker%20Hub)](https://hub.docker.com/repository/docker/regulad/amzscout-scrape)
[![Version on GitHub](https://img.shields.io/github/v/release/regulad/amzscout-scrape?include_prereleases&label=GitHub)](https://github.com/regulad/amzscout-scrape/releases)
[![Version on PyPi](https://img.shields.io/pypi/v/amzscoutscrape)](https://pypi.org/project/amzscoutscrape)
[![Version on Conda-Forge](https://img.shields.io/conda/vn/conda-forge/amzscout-scrape?label=Conda-Forge)](https://anaconda.org/conda-forge/amzscout-scrape)
[![Documentation status](https://readthedocs.org/projects/amzscout-scrape/badge)](https://amzscout-scrape.readthedocs.io/en/stable)
[![Build (GitHub Actions)](https://img.shields.io/github/workflow/status/regulad/amzscout-scrape/Build%20&%20test?label=Build%20&%20test)](https://github.com/regulad/amzscout-scrape/actions)
[![Build (Travis)](https://img.shields.io/travis/regulad/amzscout-scrape?label=Travis)](https://travis-ci.com/regulad/amzscout-scrape)
[![Build (Azure)](https://img.shields.io/azure-devops/build/regulad/<<key>>/<<defid>>?label=Azure)](https://dev.azure.com/regulad/amzscout-scrape/_build?definitionId=1&_a=summary)
[![Build (Scrutinizer)](https://scrutinizer-ci.com/g/regulad/amzscout-scrape/badges/build.png?b=main)](https://scrutinizer-ci.com/g/regulad/amzscout-scrape/build-status/main)
[![Test coverage (coveralls)](https://coveralls.io/repos/github/regulad/amzscout-scrape/badge.svg?branch=main&service=github)](https://coveralls.io/github/regulad/amzscout-scrape?branch=main)
[![Test coverage (codecov)](https://codecov.io/github/regulad/amzscout-scrape/coverage.svg)](https://codecov.io/gh/regulad/amzscout-scrape)
[![Test coverage (Scrutinizer)](https://scrutinizer-ci.com/g/regulad/amzscout-scrape/badges/coverage.png?b=main)](https://scrutinizer-ci.com/g/regulad/amzscout-scrape/?branch=main)
[![Maintainability (Code Climate)](https://api.codeclimate.com/v1/badges/<<apikey>>/maintainability)](https://codeclimate.com/github/regulad/amzscout-scrape/maintainability)
[![CodeFactor](https://www.codefactor.io/repository/github/dmyersturnbull/tyrannosaurus/badge)](https://www.codefactor.io/repository/github/dmyersturnbull/tyrannosaurus)
[![Code Quality (Scrutinizer)](https://scrutinizer-ci.com/g/regulad/amzscout-scrape/badges/quality-score.png?b=main)](https://scrutinizer-ci.com/g/regulad/amzscout-scrape/?branch=main)
[![Created with Tyrannosaurus](https://img.shields.io/badge/Created_with-Tyrannosaurus-0000ff.svg)](https://github.com/dmyersturnbull/tyrannosaurus)

Generate a dataset ready for data science from AMZScout without paying a dime.

And it’s further described in this paragraph.
[See the docs 📚](https://amzscout-scrape.readthedocs.io/en/stable/) for more info.

# About

This uses undetected chromedriver and the [AMZScout extension `njopapoodmifmcogpingplfphojnfeea`](https://chrome.google.com/webstore/detail/amazon-product-finder-amz/njopapoodmifmcogpingplfphojnfeea?utm_source=webapp&utm_medium=amzscout_wa&utm_campaign=topbanner?utm_source=webapp) to extract information from Amazon product pages.

# Using

## Dependencies

### Chrome or Chromium

#### Windows

```bash
winget install --id Google.Chrome
```

#### Ubuntu & Debian

```bash
sudo apt-get install chromium-browser
```

#### MacOS & Other

Download [here](https://www.google.com/chrome/) like anywhere else

### `chromedriver`

Make sure you install `chromedriver` on your system alongside a recent version of Chrome or Chromium.

#### brew (macOS only)

```bash
brew install chromedriver
```

#### Scoop (Windows)

Note: scoop only has chromedriver for win32 platforms. If you're on a 64-bit system, your mileage may vary.

```bash
scoop install chromedriver
```

#### Other platforms

Refer to https://sites.google.com/chromium.org/driver/downloads?authuser=0 for instructions downloading chromedriver on your respective system.

### poetry

With a good version of Python installed, run the following:

```bash
pip install pipx
pipx install poetry
```

If you get a memo about it not being on your path, Google will be your best friend.

## Installation

```bash
git clone https://github.com/regulad/amzscout-scrape.git
cd amzscout-scrape
poetry install
```

## Usage

Usage is about as simple as it gets.

```bash
poetry run amzscout-scrape --help
```

It will keep you updated, although it will take a hot minute.

Once it's done, you'll have a CSV file in the current directory.

## Proxy

Included is a simple Tailscale configuration that serves a SOCKS5 proxy on your local machine.

Copy the contents of `proxy.env-example` into `.env`, populate the fields and then perform the following command on a machine that has docker installed:

```bash
docker-compose up -d
```

Then, you can use the `--proxy` flag to use the proxy.

```bash
poetry run amzscout-scrape --proxy socks5://localhost:1055
```

# Appendix

Licensed under the terms of the [Apache License 2.0](https://spdx.org/licenses/Apache-2.0.html).
[New issues](https://github.com/regulad/amzscout-scrape/issues) and pull requests are welcome.
Please refer to the [contributing guide](https://github.com/regulad/amzscout-scrape/blob/main/CONTRIBUTING.md)
and [security policy](https://github.com/regulad/amzscout-scrape/blob/main/SECURITY.md).
Generated with [Tyrannosaurus](https://github.com/dmyersturnbull/tyrannosaurus).


CHANGELOG.md
# Changelog for amzscout-scrape

Adheres to [Semantic Versioning 2.0](https://semver.org/spec/v2.0.0.html),
[Keep a Changelog 1.0](https://keepachangelog.com/en/1.0.0/),
and [Conventional Commits 1.0](https://www.conventionalcommits.org/en/v1.0.0/).

## [0.1.0] - 2023-06-16

### Added:

- Stub files.


tests/__init__.py
"""
Utilities for tests.

Original source: https://github.com/dmyersturnbull/tyrannosaurus
Copyright 2020–2021 Douglas Myers-Turnbull
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
"""
# NOTE: If you modify this file, you should indicate your license and copyright as well.
from __future__ import annotations

import contextlib
import io
import logging
import os
import random
import shutil
import stat
import tempfile
import time
from datetime import datetime
from pathlib import Path, PurePath
from typing import Generator, Union
from warnings import warn

# Keeps created temp files; turn on for debugging
KEEP = False
# Separate logging in the main package vs. inside test functions
logger_name = Path(__file__).parent.parent.name.upper() + ".TEST"
_logger = logging.getLogger(logger_name)


class Capture(contextlib.ExitStack):
    def __init__(self):
        super().__init__()
        self._stdout = io.StringIO()
        self._stderr = io.StringIO()

    @property
    def stdout(self) -> str:
        return self._stdout.getvalue()

    @property
    def stderr(self) -> str:
        return self._stderr.getvalue()

    def __enter__(self):
        _logger.info("Capturing stdout and stderr")
        super().__enter__()
        self._stdout_context = self.enter_context(contextlib.redirect_stdout(self._stdout))
        # If the next line failed, the stdout context wouldn't exit
        # But this line is very unlikely to fail in practice
        self._stderr_context = self.enter_context(contextlib.redirect_stderr(self._stderr))
        return self

    def __exit__(self, *exc):
        _logger.info("Finished capturing stdout and stderr")
        # The ExitStack handles everything
        super().__exit__(*exc)


class TestResources:
    """
    A static singleton with utilities for filesystem operations in tests.
    Use ``TestResources.resource`` to get a file under ``tests/resources/``.

    Initializes a temporary directory with ``tempfile.TemporaryDirectory``
    and populates it with a single subdirectory, ``TestResources.global_temp_dir``.
    Temp directories for independent tests can be created underneath using
    ``TestResources.temp_dir``.
    """

    logger = _logger

    _start_dt = datetime.now()
    _start_ns = time.monotonic_ns()
    if KEEP:
        _tempfile_dir = None
        _temp_dir = Path("tmp-test-data")
    else:
        _tempfile_dir = tempfile.TemporaryDirectory()
        _temp_dir = Path(_tempfile_dir.name)

    logger.info(f"Set up main temp dir {_temp_dir.absolute()}")

    @classmethod
    @contextlib.contextmanager
    def capture(cls) -> Capture:
        """
        Context manager that captures stdout and stderr in a ``Capture`` object that contains both.
        Useful for testing code that prints to stdout and/or stderr.

        Yields:
            A ``Capture`` instance, which contains ``.stdout`` and ``.stderr``
        """
        with Capture() as cap:
            yield cap

    @classmethod
    def resource(cls, *nodes: Union[PurePath, str]) -> Path:
        """
        Gets a path of a test resource file under ``resources/``.

        Arguments:
            nodes: Path nodes under the ``resources/`` dir

        Returns:
            The Path ``resources``/``<node-1>``/``<node-2>``/.../``<node-n>``
        """
        return Path(Path(__file__).parent, "resources", *nodes).resolve()

    @classmethod
    @contextlib.contextmanager
    def temp_dir(
        cls,
        copy_resource: Union[None, str, Path] = None,
        force_delete: bool = True,
    ) -> Generator[Path, None, None]:
        """
        Context manager.
        Creates a new temporary directory underneath ``global_temp_dir``.
        Note that it deletes the directory if it already exists,
        then deletes (if the path exists) when the context closes.

        Arguments:
            copy_resource: Copy from underneath the resource dir
            force_delete: If necessary, change the permissions to delete

        Yields:
            The created directory as a ``pathlib.Path``
        """
        path = TestResources._temp_dir / ("%0x" % random.getrandbits(64))
        if path.exists():
            cls._delete_tree(path, surefire=force_delete)
        if copy_resource is None:
            path.mkdir(parents=True)
            TestResources.logger.info(f"Created empty temp dir {path}")
        else:
            path.parent.mkdir(parents=True, exist_ok=True)
            copy_path = Path(Path(__file__).parent) / "resources" / copy_resource
            shutil.copytree(str(copy_path), str(path))
            TestResources.logger.info(f"Copied {copy_resource} to temp {path}")
        yield path
        # note the global dir is only cleaned by tempfile on exit
        if not KEEP:
            cls._delete_tree(path, surefire=force_delete)

    @classmethod
    def global_temp_dir(cls) -> Path:
        """
        The global temporary directory, which is underneath ``tempfile.TemporaryDirectory``.
        The parent directory will be destroyed, along with all of its components,
        as specified by ``tempfile``.
        """
        return cls._temp_dir

    @classmethod
    def start_datetime(cls) -> datetime:
        """
        The datetime that ``tests/__init__.py`` was imported.
        """
        return cls._start_dt

    @classmethod
    def start_monotonic_ns(cls) -> int:
        """
        The nanosecond value of the ``time.monotonic`` clock at which ``tests/__init__.py`` was imported.
        """
        return cls._start_ns

    @classmethod
    def destroy(cls) -> None:
        """
        Deletes the full tempdir tree.
        """
        if not KEEP:
            cls._tempfile_dir.cleanup()

    @classmethod
    def _delete_tree(cls, path: Path, surefire: bool = False) -> None:
        def on_rm_error(func, pth, exc_info):
            # from: https://stackoverflow.com/questions/4829043/how-to-remove-read-only-attrib-directory-with-python-in-windows
            os.chmod(pth, stat.S_IWRITE)
            os.unlink(pth)

        kwargs = dict(onerror=on_rm_error) if surefire else {}
        if path.exists():
            try:
                shutil.rmtree(str(path), **kwargs)
            except OSError:
                warn(f"Testing dir {path} could not be deleted")


__all__ = ["TestResources"]


tests/test_cli.py
"""
Tests for command-line interface.

Copyright 2023 Parker Wahle <regulad@regulad.xyz>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied. See the License for the specific language governing
permissions and limitations under the License.

"""
import contextlib
import io

import pytest

from amzscoutscrape import cli

from . import TestResources


class TestCli:
    def test_cli(self):
        with TestResources.capture() as capture:
            response = cli.info()
        assert f"Processed 100 things." in capture.stdout
        assert capture.stderr.strip() == ""


if __name__ == "__main__":
    pytest.main()


amzscoutscrape/cli.py
"""
Command-line interface for amzscout-scrape.

Copyright 2023 Parker Wahle <regulad@regulad.xyz>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied. See the License for the specific language governing
permissions and limitations under the License.

"""

from __future__ import annotations

import csv
import logging
import os
import time
from pathlib import Path
from typing import Optional, cast

import typer
from _csv import Writer
from rich.logging import RichHandler
from rich.progress import track
from selenium.webdriver.remote.webdriver import WebDriver

from . import AmzscoutscrapeAssets, __copyright__, __title__, __version__, metadata
from .driver import Driver, create_fresh_driver
from .scrape import search_and_write_amazon, search_and_write_amzscout

logger = logging.getLogger(__package__)
cli = typer.Typer()

USES_OF_EXTENSION = 15
USES_OF_DEDICATED = 5


def info(n_seconds: float = 0.01, verbose: bool = False) -> None:
    """
    Get info about amzscout-scrape.

    Args:
        n_seconds: Number of seconds to wait between processing.
        verbose: Output more info

    Example:
        To call this, run: ::

            from testme import info
            info(0.02)
    """
    typer.echo(f"{__title__} version {__version__}, {__copyright__}")
    if verbose:
        typer.echo(str(metadata.__dict__))
    total = 0
    with typer.progressbar(range(100)) as progress:
        for value in progress:
            time.sleep(n_seconds)
            total += 1
    typer.echo(f"Processed {total} things.")


@cli.command()
def generate(
    filename: str = "amzscout.csv",
    verbosity: int = 0,
    headful: bool = False,
    driver_type: str = "default",
    queries: int = -1,
    skip: int = 0,
    timeout: Optional[float] = None,
    proxy: Optional[str] = None,
    extension: bool = True,
) -> None:
    """
    Generate a basic csv from AMZScout data.
    Returns:


    Args:
        skip: How many queries to skip ahead
        verbosity: How verbose the program should be. 0 is default (errors), 1 is warnings, 2 is info, 3 is debug.
        queries: The number of queries to run. Defaults to None, which means all queries.
        filename: The filename to write to. Defaults to "amzscout.csv".
        headful: Weather or not a Chrome window should be opened. This is only useful for debugging.
        driver_type: The driver to use. Defaults to "default", which is the best match for your OS. Options include "chrome", "edge", "firefox", and "undetected".
        timeout: The number of seconds to wait for the page to load before giving up.
        proxy: A proxy to use. If left unspecified, the system proxy will be utilized. If set to "direct://" no proxy will be used.
        extension: Use the legacy scraper. This is slower, but more reliable.
    """
    log_level = logging.ERROR
    match verbosity:
        case 0:
            log_level = logging.ERROR
        case 1:
            log_level = logging.WARNING
        case 2:
            log_level = logging.INFO
        case 3:
            log_level = logging.DEBUG

    logging.basicConfig(level=log_level, handlers=[RichHandler()])

    driver_enum_value: Driver = Driver.U_CHROME
    match driver_type:
        case "default":
            if extension:
                driver_enum_value = (
                    Driver.U_CHROME
                )  # safety on Amazon.com they have a lot of bot detection
            else:
                if os.name == "nt":
                    # Windows users already have Edge installed, so we'll use that.
                    driver_enum_value = Driver.EDGE
                else:
                    # Linux and macOS users probably want Firefox.
                    driver_enum_value = Driver.FIREFOX
        case "chrome":
            driver_enum_value = Driver.CHROME
        case "edge":
            driver_enum_value = Driver.EDGE
        case "firefox":
            driver_enum_value = Driver.FIREFOX
        case "undetected":
            driver_enum_value = Driver.U_CHROME
        case _:
            logger.warning(f"Invalid driver {driver_type!r}")

    with AmzscoutscrapeAssets.path("amazon_products.txt").open("r", encoding="utf-8") as fp:
        potential_queries = [line.strip() for line in fp.readlines()][skip:queries]

    filepath = Path(filename).absolute()

    exists = filepath.exists()
    with filepath.open("w" if not exists else "a", newline="", encoding="utf-8") as fp:
        csv_writer = cast(Writer, csv.writer(fp, dialect="excel"))

        typer.echo(f"Writing to {filepath.absolute()}")

        driver: WebDriver | None = None

        try:
            fails = 0
            for i, query in track(
                enumerate(potential_queries),
                description="Scraping (this WILL take a while)...",
                total=len(potential_queries),
            ):
                # Every 14 queries, restart the browser to avoid getting blocked out.
                use_count = USES_OF_DEDICATED if not extension else USES_OF_EXTENSION
                if i % use_count == 0 and driver is not None:
                    logger.info("Driver expired, killing...")
                    driver.quit()
                    driver = None
                while driver is None:
                    logger.info("Attempting to create a new driver...")
                    driver = create_fresh_driver(
                        headless=not headful,
                        timeout=timeout,
                        driver_type=driver_enum_value,
                        proxy=proxy,
                        load_extension=extension,
                    )
                try:
                    logger.info(f"Starting {query!r}, #{i + skip}...")
                    if extension:
                        search_and_write_amazon(
                            driver,
                            csv_writer,
                            query,
                            write_headers=i == 0 and not exists,
                            proxy=proxy,
                        )
                    else:
                        search_and_write_amzscout(
                            driver,
                            csv_writer,
                            query,
                            write_headers=i == 0 and not exists,
                            proxy=proxy,
                        )
                except Exception as e:
                    fails += 1
                    logger.exception(f"Error while processing query {query!r}: {e}")
                    logger.info(f"Skipping {query!r}, {fails} fails so far...")
            logger.info(
                f"Completed lookup of {len(potential_queries)} queries." f" {fails} failed."
            )
            logger.info(f"Fail rate: {fails / len(potential_queries) * 100:.2f}%")
        finally:
            if driver is not None:
                logger.info("Closing driver...")
                driver.quit()

        typer.echo("Done! Enjoy your freshly-picked data!")


if __name__ == "__main__":
    cli()


amzscoutscrape/email.py
"""
Email faking code for amzscout-scrape.

Copyright 2023 Parker Wahle <regulad@regulad.xyz>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied. See the License for the specific language governing
permissions and limitations under the License.

"""
from time import sleep
from uuid import uuid4

from selenium.webdriver.chrome.webdriver import WebDriver
from selenium.webdriver.common.by import By


def get_writeonly_tempemail(driver: WebDriver) -> str:
    """
    Get a clean email using selenium that shouldn't already have an account attached
    """
    old_tab = driver.current_window_handle

    driver.switch_to.new_window("tab")
    driver.get("https://temp-mail.org/en/")

    while "Loading" in driver.find_element(By.ID, "mail").get_attribute("value"):
        sleep(0.1)

    email = driver.find_element(By.ID, "mail").get_attribute("value")

    driver.close()
    driver.switch_to.window(old_tab)

    return email


def get_readable_tempemail(driver: WebDriver) -> str:
    """
    Use a legit email service to get a clean email
    """
    raise NotImplementedError


def get_random_plausible_email(driver: WebDriver, *, domain: str = "gmail.com") -> str:
    """
    Get a random clean email that shouldn't already have an account attached
    """
    # Slight problem: AMZScout seems to have been detecting our scraping and is now locking out every
    # non-gmail email. It's their fault for not using email verification & a captcha!

    # I got around this by cycling my IP address, but we need some better type of proxy to avoid getting kneecapped
    # in the future. Check out my project regulad/sticky-starfish
    username = uuid4().hex
    short_username = username[: int(len(username) // 4)]
    email = f"{short_username}@{domain}"
    return email


__all__ = ["get_random_plausible_email", "get_writeonly_tempemail", "get_readable_tempemail"]


amzscoutscrape/proxy.py
"""
Proxy-juggling code for amzscout-scrape.

Copyright 2023 Parker Wahle <regulad@regulad.xyz>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied. See the License for the specific language governing
permissions and limitations under the License.

"""
import logging
from concurrent.futures import ThreadPoolExecutor
from functools import partial
from threading import Event
from typing import Sequence, cast

import requests
from requests import Session

_CACHE: Sequence[dict] | None = None
_USED = set()
logger = logging.getLogger(__package__)


def _fetch_proxies() -> Sequence[dict]:
    global _CACHE
    logger.info("Fetching proxies from proxylist.geonode.com...")
    with requests.get(
        "https://proxylist.geonode.com/api/proxy-list?limit=500&page=1&sort_by=lastChecked&sort_type=desc&country=US",
        timeout=60,
    ) as r:
        _CACHE = r.json()["data"]
    logger.info(f"Fetched {len(_CACHE)} proxies.")
    return cast(Sequence[dict], _CACHE)


def setup_proxy_for_requests(session: Session, proxy: str | None = None) -> None:
    if proxy is None:
        return

    if proxy == "direct://":
        session.trust_env = False
        return

    proxies = {
        "http": proxy,
        "https": proxy,
        "ftp": proxy,
    }
    session.proxies.update(proxies)


def fetch_working_geonode_proxy() -> str:
    """
    Get a random SOCKS5 proxy from proxylist.geonode.com.
    Returns: A SOCKS5 proxy.

    """
    global _CACHE, _USED

    proxy_found = Event()
    good_proxy: str | None = None

    with ThreadPoolExecutor(
        thread_name_prefix="ProxySearch", max_workers=64
    ) as executor, Session() as session:

        def _try_proxy(proxy_with_protocol: str) -> bool:
            nonlocal good_proxy

            logger.debug(f"Trying proxy {proxy_with_protocol}...")

            try:
                with session.get(
                    "https://api.ipify.org",
                    timeout=30,
                    proxies={"http": proxy_with_protocol, "https": proxy_with_protocol},
                ) as r:
                    if not proxy_found.is_set() and r.ok:
                        proxy_found.set()  # no need for a lock?
                        good_proxy = proxy_with_protocol
                        return True
                    else:
                        return False
            except (
                requests.exceptions.ProxyError,
                requests.exceptions.ConnectTimeout,
                requests.exceptions.ReadTimeout,
                requests.exceptions.ConnectionError,
            ) as e:
                return False

        for proxy in _CACHE or _fetch_proxies():
            if proxy["_id"] not in _USED:
                _USED.add(proxy["_id"])
            else:
                continue

            host = proxy["ip"]
            port = proxy["port"]
            for protocol in proxy["protocols"]:
                executor.submit(partial(_try_proxy, f"{protocol}://{host}:{port}"))

        proxy_found.wait()
        executor.shutdown(wait=False, cancel_futures=True)
        logger.info(f"Proxy {good_proxy} succeeded, using...")
        return good_proxy


def ip_of(proxy: str) -> str:
    with Session() as s:
        setup_proxy_for_requests(s, proxy)
        with s.get("https://api.ipify.org") as r:
            return r.text


__all__ = ("fetch_working_geonode_proxy", "setup_proxy_for_requests", "ip_of")


amzscoutscrape/utils.py
"""
Utility code for amzscout-scrape.

Copyright 2023 Parker Wahle <regulad@regulad.xyz>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied. See the License for the specific language governing
permissions and limitations under the License.

"""
import functools
import logging
import time
import warnings
from typing import Mapping, TypeVar

K = TypeVar("K")
V = TypeVar("V")

logger = logging.getLogger(__package__)


def reverse_map(dictionary: Mapping[K, V], value: V) -> K | None:
    for k, v in dictionary.items():
        if v == value:
            return k


def deprecated(func):
    """This is a decorator which can be used to mark functions
    as deprecated. It will result in a warning being emitted
    when the function is used."""

    @functools.wraps(func)
    def new_func(*args, **kwargs):
        warnings.simplefilter("always", DeprecationWarning)  # turn off filter
        warnings.warn(
            "Call to deprecated function {}.".format(func.__name__),
            category=DeprecationWarning,
            stacklevel=2,
        )
        warnings.simplefilter("default", DeprecationWarning)  # reset filter
        return func(*args, **kwargs)

    return new_func


# https://keestalkstech.com/2021/03/python-utility-function-retry-with-exponential-backoff/
# with special sauce mods
def retry(tries=5, backoff_seconds=1):
    def rwb(f):
        @functools.wraps(f)
        def wrapper(*args, **kwargs):
            attempts = 0
            while True:
                try:
                    return f(*args, **kwargs)
                except Exception as e:
                    if (
                        type(e) == KeyboardInterrupt
                        or type(e) == SystemExit
                        or type(e) == NotImplementedError
                    ):
                        raise

                    if attempts == tries:
                        raise

                    sleep = backoff_seconds * 2**attempts

                    logger.exception(f"Retrying {f.__name__} in {sleep} seconds due to {e}")

                    time.sleep(sleep)
                    attempts += 1

        return wrapper

    return rwb


# TODO: add timeout context manager
# https://stackoverflow.com/questions/2281850/timeout-function-if-it-takes-too-long-to-finish


__all__ = ("reverse_map", "deprecated", "retry")


amzscoutscrape/__init__.py
"""
Metadata for this amzscout-scrape.

Copyright 2023 Parker Wahle <regulad@regulad.xyz>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied. See the License for the specific language governing
permissions and limitations under the License.

"""
import logging
from importlib.metadata import PackageNotFoundError
from importlib.metadata import metadata as __load
from pathlib import Path
from typing import Sequence

pkg = Path(__file__).absolute().parent.name
logger = logging.getLogger(pkg)
metadata = None
try:
    metadata = __load(pkg)
    __status__ = "Development"
    __copyright__ = "Copyright 2023"
    __date__ = "2023-06-16"
    __uri__ = metadata["home-page"]
    __title__ = metadata["name"]
    __summary__ = metadata["summary"]
    __license__ = metadata["license"]
    __version__ = metadata["version"]
    __author__ = metadata["author"]
    __maintainer__ = metadata["maintainer"]
    __contact__ = metadata["maintainer"]
except PackageNotFoundError:  # pragma: no cover
    logger.error(f"Could not load package metadata for {pkg}. Is it installed?")


class AmzscoutscrapeAssets:
    """
    Utilities and resources for ${Project}.
    """

    @classmethod
    def path(cls, *nodes: Sequence[str]) -> Path:
        """
        Gets the ``Path`` of an asset under ``resources``.

        Args:
            nodes: Path nodes underneath ``resources``

        Returns:
            The final ``Path``

        Raises:
            FileNotFoundError: If the path does not exist
        """
        path = Path(Path(__file__).parent, "resources", *nodes)
        if not path.exists():
            raise FileNotFoundError(f"Asset {path} not found")
        return path


if __name__ == "__main__":  # pragma: no cover
    if metadata is not None:
        print(f"{pkg} (v{metadata['version']})")
    else:
        print(f"Unknown project info for {pkg}")
